src/vmm/src/devices/virtio/vhost_user.rs

```diff
@@ -0,0 +1,259 @@
+use vhost::{Error as VhostError, VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use vm_memory::{Address, Error as MmapError, GuestMemory, GuestMemoryError, GuestMemoryRegion};
+
+use crate::devices::virtio::queue::Queue;
+use crate::devices::virtio::IrqTrigger;
+use crate::vstate::memory::GuestMemoryMmap;
+
+/// vhost-user error.
+#[derive(Error, Debug, displaydoc::Display)]
+pub enum VhostUserError {
+    /// Invalid available address
+    AvailAddress(GuestMemoryError),
+    /// Failed to connect to UDS Unix stream: {0}
+    Connect(#[from] std::io::Error),
+    /// Invalid descriptor table address
+    DescriptorTableAddress(GuestMemoryError),
+    /// Get features failed: {0}
+    VhostUserGetFeatures(VhostError),
+    /// Get protocol features failed: {0}
+    VhostUserGetProtocolFeatures(VhostError),
+    /// Set owner failed: {0}
+    VhostUserSetOwner(VhostError),
+    /// Set features failed: {0}
+    VhostUserSetFeatures(VhostError),
+    /// Set protocol features failed: {0}
+    VhostUserSetProtocolFeatures(VhostError),
+    /// Set mem table failed: {0}
+    VhostUserSetMemTable(VhostError),
+    /// Set vring num failed: {0}
+    VhostUserSetVringNum(VhostError),
+    /// Set vring addr failed: {0}
+    VhostUserSetVringAddr(VhostError),
+    /// Set vring base failed: {0}
+    VhostUserSetVringBase(VhostError),
+    /// Set vring call failed: {0}
+    VhostUserSetVringCall(VhostError),
+    /// Set vring kick failed: {0}
+    VhostUserSetVringKick(VhostError),
+    /// Set vring enable failed: {0}
+    VhostUserSetVringEnable(VhostError),
+    /// Failed to read vhost eventfd: {0}
+    VhostUserMemoryRegion(MmapError),
+    /// Invalid used address
+    UsedAddress(GuestMemoryError),
+}
+
+/// vhost-user socket handle
+#[derive(Clone)]
+pub struct VhostUserHandle {
+    pub vu: Master,
+    pub socket_path: String,
+}
+
+impl std::fmt::Debug for VhostUserHandle {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("VhostUserHandle")
+            .field("socket_path", &self.socket_path)
+            .finish()
+    }
+}
+
+impl VhostUserHandle {
+    /// Connect to the vhost-user backend socket and mark self as an
+    /// owner of the session.
+    pub fn new(socket_path: &str, num_queues: u64) -> Result<Self, VhostUserError> {
+        let stream = UnixStream::connect(socket_path).map_err(VhostUserError::Connect)?;
+
+        let vu = Master::from_stream(stream, num_queues);
+        vu.set_owner().map_err(VhostUserError::VhostUserSetOwner)?;
+
+        Ok(Self {
+            vu,
+            socket_path: socket_path.to_string(),
+        })
+    }
+
+    pub fn set_features(&self, features: u64) -> Result<(), VhostUserError> {
+        self.vu
+            .set_features(features)
+            .map_err(VhostUserError::VhostUserSetFeatures)
+    }
+
+    /// Set vhost-user protocol features to the backend.
+    pub fn set_protocol_features(
```

Comment:
Actually this one seems to be not used anywhere in current implementation. It was used in the original implementation in restore path, but because currently we don't support save/restore of vhost devices, it is not used. But in the future it will be and will need to be public.